File: 22. generateParenthesis.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<String>();
        backTrack(res, new StringBuilder(), 0, 0, n);
        return res;
    }
    public void backTrack(List<String>res, StringBuilder cur, int open, int close, int max){
        if (cur.length() == max*2) {
            res.add(cur.toString());
            return;
        }
        if (open < max){
            cur.append("(");
            backTrack(res, cur, open+1, close, max);
            cur.deleteCharAt(cur.length()-1);
        }
        
        if (close < open) {
            cur.append(")");
            backTrack(res, cur, open, close+1, max);
            
            cur.deleteCharAt(cur.length() - 1);
        }
        
    }
}

File: 22. generateParenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def backTrack(S=[], left = 0, right = 0):
            if len(S) == 2*n:
                res.append("".join(S))
            if left < n:
                S.append("(")
                backTrack(S, left+1, right);
                S.pop()
                print(S == A)
            if right < left:
                S.append(")")
                backTrack(S, left, right+1)
                S.pop()
        backTrack()
        return res